src/d2lfetch-dedupe.js

```diff
@@ -1,51 +1,143 @@
+ *
+ * @typedef {Record<string, InflightRequestEntry>} InflightRequestInfo
+ */
+
 export class D2LFetchDedupe {
 
 	constructor() {
-		this._inflightRequests = this._inflightRequests || [];
+		/** @type {InflightRequestInfo} */
+		this._inflightRequests = {};
+		this._nextReqId = 0;
 	}
 
+	/**
+	 * @param {Request} request
+	 * @param {MiddlewareFunc} next
+	 *
+	 * @returns {Promise<Response>}
+	 */
 	dedupe(request, next) {
 		if (false === request instanceof Request) {
 			return Promise.reject(new TypeError('Invalid request argument supplied; must be a valid window.Request object.'));
 		}
 
 		if (request.method !== 'GET' && request.method !== 'HEAD' && request.method !== 'OPTIONS') {
-			if (!next) {
-				return Promise.resolve(request);
-			}
 			return next(request);
 		}
 
 		const key = this._getKey(request);
-		if (this._inflightRequests[key]) {
-			this._inflightRequests[key].count++;
-			return this._inflightRequests[key].action;
-		}
+		const reqId = this._nextReqId++;
+
+		let newRequest = false;
+
+		if (!this._inflightRequests[key]) {
+			const abortController = window.AbortController
+				? new AbortController()
+				: null;
+
+			const inflightRequest = new Request(request, {
+				signal: (abortController || {}).signal
+			});
 
-		if (!next) {
-			return Promise.resolve(request);
+			this._inflightRequests[key] = {
+				inflightRequest,
+				abortController,
+				dedupedRequests: {}
+			};
+
+			newRequest = true;
 		}
 
-		const result = next(request);
-		if (result && result instanceof Promise) {
-			this._inflightRequests[key] = { count: 1 };
-			this._inflightRequests[key].action = result
-				.then(function(response) {
-					const usedMultiple = this._inflightRequests[key].count !== 1;
-					delete this._inflightRequests[key];
-					return this._clone(response, usedMultiple);
-				}.bind(this))
-				.catch(function(err) {
-					delete this._inflightRequests[key];
-					return Promise.reject(err);
-				}.bind(this));
-
-			return this._inflightRequests[key].action;
+		const dedupedRequest = new Promise((resolve, reject) => {
+			const inflightRequestEntry = this._inflightRequests[key];
+
+			inflightRequestEntry.dedupedRequests[reqId] = {
+				resolvers: { resolve, reject }
+			};
+
+			if (request.signal && typeof request.signal.addEventListener === 'function') {
+				/**
+				 * @param {AbortSignal} signal
+				 * @param {string} key
+				 * @param {number} reqId
+				 */
+				const addAbortEventListener = (signal, key, reqId) => {
+					const handler = () => {
+						signal.removeEventListener('abort', handler);
+
+						const inflightRequestEntry = this._inflightRequests[key];
+
+						if (!inflightRequestEntry) {
+							return;
+						}
+
+						if (Object.keys(inflightRequestEntry.dedupedRequests).length === 1) {
+							const abortController = inflightRequestEntry.abortController;
+
+							if (abortController && typeof abortController.abort === 'function') {
```

Comment:
if `abortController` isn't `null` we know it's an `AbortController`. I don't think we need the `abortController.abort` check.
```suggestion
							if (abortController) {
```